`load_commondata`, `level0_commondata_wc`, and `make_level1_data`

FrancescoMerlotti opened the pull request:
1. added `load_commondata` to `validphys2/src/validphys/core.py`
2. added `level0_commondata_wc` and `make_level1_data` to `validphys2/src/validphys/pseudodata.py`
to use `chi2_provider.py` which needs the two functions added in `validphys2/src/validphys/pseudodata.py`

Note: in `validphys2/src/validphys/pseudodata.py` function `make_replica` already takes into account the right multiplicative errors since they are stored as percentage in `systematics_table`, they are converted one the spot to fractions, and it is computed the product with the pseudodata which are made with `level0_commondata_wc`

comane commented on validphys2/src/validphys/pseudodata.py:
```diff
@@ -235,6 +237,154 @@ def indexed_make_replica(groups_index, make_replica):
+
+    level0_commondata_instances_wc = []
+
+    # argument fakepdf is passed as str
+    # generate a core.PDF object with name equal to fakepdf argument
+    fakepdf = PDF(name=fakepdf)
```
This could be done by changing the input instead.
I don't think that it makes sense in general for `fakepdf` to be a string, it might be better for it to be a `core.PDF` object.
To do so you can have a look at the parse rules in `validphys.config.py` module.
In particular, if I am not mistaken, it should be already implemented that way within nnpdf.

Note, however, that this has to be tested, as there might be some parts of the SIMUnet code that still assume that `fakepdf` is a string

FrancescoMerlotti replied:
I added `parse_fakepdf` to `config.py`. It does what we want, _i.e._ it parse the fakepdf in the runcard as a `Core.PDF` object, I have to test in other scenarios, _e.g._ closuretests

comane reviewed (commented):
Thank you for this.
Could you please also add the chi2_provider functions?
You could add them, for instance, in the `simunet_analysis.py` module

FrancescoMerlotti commented:
Added the `chi2_provider.py` functions to `pseudodata.py`.
I guess that `compute_chi2` and `write_chi2` should be move in another file among the providers.
I also added some examples for some of the implemented functions, I am working on the others right now

comane commented on validphys2/src/validphys/pseudodata.py:
```diff
@@ -235,6 +242,402 @@ def indexed_make_replica(groups_index, make_replica):
+    return sm_dict
+
+
+def load_contamination(
+        contamination_parameters,
+        theoryid,
+        dataset_inputs
```
Could you please move this to another place? Perhaps `simunet_analysis.py`?

comane commented on validphys2/src/validphys/pseudodata.py:
```diff
@@ -235,6 +242,402 @@ def indexed_make_replica(groups_index, make_replica):
+    return bsm_dict
+
+
+def compute_chi2(
+        make_level1_list_data,
+        sm_predictions,
+        groups_covmat,
+        load_contamination
```
Also this should be moved somewhere else (`simunet_analysis.py`) and have another name that is more informative, given that this function can only be used for very specific cases and is not used to compute the chi2 in general

comane commented on validphys2/src/validphys/pseudodata.py:
```diff
@@ -235,6 +242,402 @@ def indexed_make_replica(groups_index, make_replica):
+    return chi2_dict
+
+
+def write_chi2(
+        pdf,
+        compute_chi2,
+        level0_commondata_wc
+    ):
+
```
Same here as above

comane commented on validphys2/src/validphys/coredata.py:
```diff
@@ -263,6 +263,10 @@ def additive_errors(self):
         add_table.columns = add_systype["name"].to_numpy()
         return add_table.loc[:, add_table.columns != "SKIP"]
 
+    @property
+    def cuts(self):
+        indices = self.commondata_table.index
+        return slice(indices[0]-1, indices[-1])
```
Is this used anywhere?

FrancescoMerlotti replied:
I left a comment at line 2085 in `validphys.simunet_analysis.py` where it is used

FrancescoMerlotti commented on validphys2/src/validphys/simunet_analysis.py:
```diff
@@ -1969,4 +1971,168 @@ def principal_component_vectors(fisher_information_matrix, simu_parameters_names
+            if bsm_fac.shape[0] == 1:
+                data_values = dataset.central_values * bsm_fac
+            else:
+                cuts = dataset.cuts
+                data_values = dataset.central_values * bsm_fac[cuts]
```
It is used here
